lib/puppet-lint/plugins/check_parameter_documentation.rb

```diff
@@ -25,51 +23,76 @@ def check
         end
       end
 
-      next if is_private
-
-      params = []
-      e = idx[:param_tokens].each
-      begin
-        while (ptok = e.next)
-          if ptok.type == :VARIABLE
-            params << ptok
-            nesting = 0
-            # skip to the next parameter to avoid finding default values of variables
-            while true
-              ptok = e.next
-              case ptok.type
-              when :LPAREN
-                nesting += 1
-              when :RPAREN
-                nesting -= 1
-              when :COMMA
-                break unless nesting > 0
-              end
-            end
-          end
-        end
-      rescue StopIteration; end
+      params = extract_params(idx)
 
       # warn about duplicates
       doc_params_duplicates.each do |parameter, tokens|
         tokens.each do |token|
            notify :warning, {
-             :message => "Duplicate parameter documentation for #{parameter}",
+             :message => "Duplicate #{type_str(idx)} parameter documentation for #{parameter}",
              :line    => token.line,
              :column  => token.column + token.value.match(/\A\s*(@param\s*)?/)[0].length + 1 # `+ 1` is to account for length of the `#` COMMENT token.
            }
         end
       end
 
+      # warn about documentation for parameters that don't exist
+      doc_params.each do |parameter, token|
+        next if params.find { |p| p.value == parameter }
+
+        notify :warning, {
+          :message => "No matching #{type_str(idx)} parameter for documentation of #{parameter}",
+          :line    => token.line,
+          :column  => token.column + token.value.match(/\A\s*(@param\s*)?/)[0].length + 1
+        }
+      end
+
+      # Warn about missing documentation unless class/type is marked private
+      next if is_private
+
       params.each do |p|
         next if doc_params.has_key? p.value
-        idx_type = idx[:type] == :CLASS ? "class" : "defined type"
         notify :warning, {
-          :message => "missing documentation for #{idx_type} parameter #{idx[:name_token].value}::#{p.value}",
+          :message => "missing documentation for #{type_str(idx)} parameter #{idx[:name_token].value}::#{p.value}",
           :line    => p.line,
           :column  => p.column,
         }
       end
     end
   end
+
+  private
+
+  def type_str(idx)
+    idx[:type] == :CLASS ? "class" : "defined type"
+  end
+
+  def extract_params(idx)
+    return [] if idx[:param_tokens].nil?
+
+    params = []
+    e = idx[:param_tokens].each
+    begin
+      while (ptok = e.next)
+        if ptok.type == :VARIABLE
+          params << ptok
+          nesting = 0
+          # skip to the next parameter to avoid finding default values of variables
+          while true
+            ptok = e.next
+            case ptok.type
+            when :LPAREN
+              nesting += 1
+            when :RPAREN
+              nesting -= 1
```

Comment:
I know I wrote this code, but if trailing commas are not used, I think here you also need to break if nesting < 0

Reply:
I couldn't find a good example of when it mattered. It turns out `e.next` raises a `StopIteration`. For example, `class myclass($foo, $bar) {}` will only give `$foo`, `,` and `$bar` as tokens. I was afraid it would also continue iterating with `)` and others, but that's not the case.